File: backend/SpaceStay.Core/Dtos/Dtos.cs
using System.ComponentModel.DataAnnotations;
using SpaceStay.Core.Domain;

namespace SpaceStay.Core.Dtos;

// DTOs de entrada e saída. Os de requisição usam data annotations para validação
// (Parte 5); as entidades do EF nunca são expostas diretamente.

// Auth
public class RegisterGuestRequest
{
    [Required, StringLength(120, MinimumLength = 2)]
    public string Name { get; set; } = null!;

    [Required, EmailAddress, StringLength(180)]
    public string Email { get; set; } = null!;

    [Required, StringLength(100, MinimumLength = 6)]
    public string Password { get; set; } = null!;

    [StringLength(60)]
    public string? Nationality { get; set; }

    public bool MedicalClearance { get; set; }
}

public class LoginRequest
{
    [Required, EmailAddress]
    public string Email { get; set; } = null!;

    [Required]
    public string Password { get; set; } = null!;
}

public record AuthResponse(string Token, DateTime ExpiresAt, string UserType, string? Role, int Id, string Name);

// Módulos, sensores e leituras
public record ModuleResponse(int Id, string Name, ModuleType Type, int Capacity, ModuleStatus Status, int OpenAlerts);

public record SensorResponse(int Id, SensorType Type, string Unit, decimal MinSafe, decimal MaxSafe);

public record ReadingResponse(
    int SensorId, SensorType Type, string Unit, decimal? Value, DateTime? RecordedAt,
    decimal MinSafe, decimal MaxSafe, string Status, string Comfort);

// Reservas
public class CreateBookingRequest
{
    [Range(1, int.MaxValue, ErrorMessage = "moduleId inválido.")]
    public int ModuleId { get; set; }

    [Required] public DateTime CheckIn { get; set; }
    [Required] public DateTime CheckOut { get; set; }
}

public class UpdateBookingRequest
{
    public DateTime? CheckIn { get; set; }
    public DateTime? CheckOut { get; set; }
    public BookingStatus? Status { get; set; }
}

public record BookingResponse(
    int Id, int GuestId, string? GuestName, int ModuleId, string? ModuleName,
    DateTime CheckIn, DateTime CheckOut, BookingStatus Status);

// Alertas
public record AlertResponse(
    int Id, int SensorId, SensorType SensorType, int ModuleId, string? ModuleName,
    AlertSeverity Severity, string Message, AlertStatus Status, DateTime CreatedAt,
    int? ResolvedBy, string? ResolvedByName);

// Ingestão de leituras (ponte IoT da Parte 6)
public class CreateReadingRequest
{
    [Range(1, int.MaxValue, ErrorMessage = "sensorId inválido.")]
    public int SensorId { get; set; }

    [Required] public decimal Value { get; set; }

    // Opcional; quando ausente, usa o instante atual.
    public DateTime? RecordedAt { get; set; }
}

public record ReadingIngestResponse(long ReadingId, bool OutOfRange, bool AlertCreated, AlertResponse? Alert);

// Excursões
public record ExcursionResponse(
    int Id, string Name, string? Description, int Capacity, DateTime ScheduledAt,
    int BookedCount, int AvailableSpots, bool BookedByMe);

public record BookExcursionResponse(int ExcursionBookingId, int ExcursionId, int GuestId, ExcursionBookingStatus Status);

File: mobile/app/src/main/java/br/com/spacestay/data/Dtos.kt
package br.com.spacestay.data

// DTOs que espelham as respostas JSON da API (camelCase). Os enums chegam como String
// e são interpretados na UI (mantém o mapeamento simples).

data class LoginRequest(val email: String, val password: String)

data class RegisterRequest(
    val name: String,
    val email: String,
    val password: String,
    val nationality: String?,
    val medicalClearance: Boolean
)

data class CreateBookingRequest(
    val moduleId: Int,
    val checkIn: String,
    val checkOut: String
)

data class AuthResponse(
    val token: String,
    val expiresAt: String,
    val userType: String,   // "guest" | "staff"
    val role: String?,      // admin | engineer | concierge (equipe)
    val id: Int,
    val name: String
)

data class PagedResult<T>(
    val items: List<T> = emptyList(),
    val page: Int = 1,
    val pageSize: Int = 0,
    val total: Int = 0,
    val totalPages: Int = 0
)

data class ModuleResponse(
    val id: Int,
    val name: String,
    val type: String,       // suite | common | lab
    val capacity: Int,
    val status: String,     // available | occupied | maintenance
    val openAlerts: Int
)

data class ReadingResponse(
    val sensorId: Int,
    val type: String,       // o2 | co2 | pressure | temperature | humidity | water
    val unit: String,
    val value: Double?,
    val recordedAt: String?,
    val minSafe: Double,
    val maxSafe: Double,
    val status: String,     // safe | warning | critical | no-data
    val comfort: String     // Ideal | Atenção | Crítico | Sem dados
)

data class AlertResponse(
    val id: Int,
    val sensorId: Int,
    val sensorType: String,
    val moduleId: Int,
    val moduleName: String?,
    val severity: String,   // info | warning | critical
    val message: String,
    val status: String,     // open | acknowledged | resolved
    val createdAt: String,
    val resolvedBy: Int?,
    val resolvedByName: String?
)

data class ExcursionResponse(
    val id: Int,
    val name: String,
    val description: String?,
    val capacity: Int,
    val scheduledAt: String,
    val bookedCount: Int,
    val availableSpots: Int,
    val bookedByMe: Boolean = false
)

data class BookExcursionResponse(
    val excursionBookingId: Int,
    val excursionId: Int,
    val guestId: Int,
    val status: String
)

data class BookingResponse(
    val id: Int,
    val guestId: Int,
    val guestName: String?,
    val moduleId: Int,
    val moduleName: String?,
    val checkIn: String,
    val checkOut: String,
    val status: String      // confirmed | active | completed
)
